### ISSUE TITLE
ImageDataGenerator: image channel issue

### ISSUE THREAD

**Issue by gtuzi** [fchollet/keras#5846]
Using "ImageDataGenerator" on RGB images seems to generate images with non-RGB values. I'm having issues figuring out the mapping to throw the generated images to RGB. 

Example:

*an RGB image*
<img width="595" alt="screen shot 2017-03-18 at 1 23 07 am" src="https://cloud.githubusercontent.com/assets/3085599/24069240/94b54fc6-0b79-11e7-923b-740eac923856.png">

*an <? channels> image* generated from `ImageDataGenerator`'s `flow()`
<img width="548" alt="screen shot 2017-03-18 at 1 24 35 am" src="https://cloud.githubusercontent.com/assets/3085599/24069247/af96ce28-0b79-11e7-9d62-57f8a64ad28f.png">

I have tried varying the order of the channels but I cannot retrieve a normal image. Has anyone seen this before ?

Keras version: 2.0.1

--- 

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by joelthchao** [comment]
It seems like this is not a bug. According to your sample image from `ImageDataGenerator` 's `flow()`, there might be following operations apply on your image:
1. Random crop with `fill_mode='nearest'`
2. Normalization on pixel value

Please provide your code to see what happen

**Comment by gtuzi** [comment]
Joel, see the following declaration:

`traindatagen = ImageDataGenerator(featurewise_center=False,
                                  featurewise_std_normalization=False,
                                  rotation_range=20, width_shift_range=0.3, 
                                  height_shift_range=0.3)`

**Comment by gtuzi** [comment]
Joel, just to emphasize, I am concerned with the color channels - I am not getting an RGB image after the image transformation.

**Comment by joelthchao** [comment]
Nothing wrong happen in my test, not sure if your images saving process has any problem.
Environment:
```
ENV:
{
    "floatx": "float32",
    "backend": "tensorflow",
    "epsilon": 1e-07,
    "image_dim_ordering": "tf"
}
keras.__version__ == '2.0.1'
```
Test script:
```python
import numpy as np
from scipy.misc import imread, imsave
from keras.preprocessing.image import ImageDataGenerator

traindatagen = ImageDataGenerator(featurewise_center=False, featurewise_std_normalization=False, rotation_range=20, width_shift_range=0.3, height_shift_range=0.3)

image = imread('test_image.jpg')
images = image[np.newaxis, :, :, :]

process_images = traindatagen.flow(images).next()
process_image = process_images[0, :, :, :]
imsave('test_process_image.jpg', process_image)
```

**Comment by gtuzi** [comment]
This probably doesn't matter - but I'm not saving the image. The images are just loaded in memory, transformed and then plotted with `plt`.

**Comment by joelthchao** [comment]
Please provides complete code. My concern is the interpretation of channel might different between libraries.

**Comment by gtuzi** [comment]
`
import numpy as np
import matplotlib.pyplot as plt
import cv2
images = []

for line in lines:
    sourcepath = line[0]
    filename = sourcepath.split('/')[-1]
    current_path = 'data/IMG/' + filename
    image = cv2.imread(current_path)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    images.append(image)

X = np.array(images)

image = X[0].squeeze()
plt.figure(figsize=(10,10))
plt.imshow(image.squeeze())
`
Here I get:

<img width="595" alt="screen shot 2017-03-18 at 1 23 07 am" src="https://cloud.githubusercontent.com/assets/3085599/24070255/e81005d8-0b8f-11e7-88fc-19b589677a74.png">

---

Next:

`testgenerator = ImageDataGenerator(featurewise_center=False,featurewise_std_normalization=False,rotation_range=20, width_shift_range=0.3, height_shift_range=0.3)
images = image[np.newaxis, :, :, :]
testgenerator.fit(images)

image = testgenerator.flow(images).next().squeeze()
plt.figure(figsize=(10,10))
plt.imshow(image.squeeze())`

and here I get:

<img width="480" alt="screen shot 2017-03-18 at 4 04 47 am" src="https://cloud.githubusercontent.com/assets/3085599/24070262/10f9b732-0b90-11e7-9878-f37503cc4408.png">

**Comment by joelthchao** [comment]
Short script to prove that problem not comes from ImageDataGenerator
```python
import numpy as np
from keras.preprocessing.image import ImageDataGenerator

image1 = np.stack([np.zeros((4, 4)), np.ones((4, 4)), np.ones((4, 4)) * 2], axis=-1)
testgenerator = ImageDataGenerator(featurewise_center=False,featurewise_std_normalization=False,rotation_range=20, width_shift_range=0.3, height_shift_range=0.3)
images = image1[np.newaxis, :, :, :]
testgenerator.fit(images)
image2 = testgenerator.flow(images).next().squeeze()

# No channel swap happens
assert np.all(image1 == image2)
```

**Comment by joelthchao** [comment]
I understand what happen now, please change your code in the end of script to:
```python
image = testgenerator.flow(images).next().squeeze()
plt.figure(figsize=(10,10))
plt.imshow(image)

# plt interpret value by np.dtype, this will make everything back to normal
image2 = image.astype(np.unit8) 
plt.imshow(image2)
```

**Comment by gtuzi** [comment]
That was it! It must be `uint8`. I was casting to `uint16` before this post, and it was giving me the same problem.

**Comment by phisad** [comment]
Awesome! Thanks for the explanation! Wondered about this a while ago.

**Comment by anurag0406** [comment]
Where have you added uint8 exacty in your code 
Please share, I have a similar error
> I was casting to `int` and it gave me the same problem. The colors were actually inverted, if I did plt.imshow(255-img) it would look correct. Using type `unit8` fixed it though. Thanks!
